src/Coordination/KeeperServer.cpp

```diff
@@ -143,20 +195,59 @@ void KeeperServer::startup(bool enable_ipv6)
     {
         LOG_INFO(log, "No config in log store and snapshot, probably it's initial run. Will use config from .xml on disk");
     }
+}
+
+void KeeperServer::recoveryMode(nuraft::raft_params & params)
+{
+    LOG_WARNING(
+        log,
+        "This instance is in recovery mode. Until the quorum is restored, no requests should be sent to any "
+        "of the cluster instances. This instance will start accepting requests only when the recovery is finished.");
+
+    auto latest_config = state_manager->load_config();
 
+    nuraft::ptr<nuraft::cluster_config> new_config = std::make_shared<nuraft::cluster_config>(0, latest_config ? latest_config->get_log_idx() : 0);
+    new_config->set_log_idx(state_manager->load_log_store()->next_slot());
+
+    new_config->get_servers() = last_local_config->get_servers();
+
+    state_manager->save_config(*new_config);
+    params.with_custom_commit_quorum_size(1);
+    params.with_custom_election_quorum_size(1);
+}
+
+void KeeperServer::forceRecovery()
+{
+    is_recovering = true;
```

Comment:
is it intentional to set this flag before lock? If yes then add a comment, please.

Reply:
Will add.
It's notifying other write threads to stop trying to process requests, e.g. config update thread.